### PR TITLE
Portal CRUD IT uses setUp and tearDown (#2400)

### PR DESCRIPTION
**Pull request by noah-aviel-dove**
Author (check before every review)

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue
- [x] Added `DCP/1` label                                   <sub>or this PR does not target an `hca/*` branch</sub>
- [x] Created issue to track porting these changes          <sub>or this PR does not need porting</sub> 
- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>
- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in chain            <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>
- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`

Primary reviewer (before pushing merge commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Build passes in `sandbox`                             <sub>or added `no sandbox` label</sub>
- [x] Reindexed `sandbox`                                   <sub>or this PR does not require reindexing</sub>
- [x] Added PR reference to merge commit
- [x] Moved linked issue to Merged
- [ ] Pushed merge commit to Github

Primary reviewer (after pushing merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened chain                                       <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [x] Commented on demo expectations                        <sub>or labeled as `no demo`</sub>
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Reindexed `dev`                                       <sub>or this PR does not require reindexing</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Unassign reviewer from PR

### CODE REVIEW

File: test/integration_test.py
Comment (achave11) [inline test/integration_test.py:618]: Should it have a return type `PortalService ` ? Up to you.

**Review by achave11 — APPROVED** [review]
Looks good, nice work!

**Comment by noah-aviel-dove** [comment]
Integration test passes once #2399 is applied on top of this, otherwise the relevant test is skipped.

File: test/integration_test.py
Comment (hannes-ucsc) [inline test/integration_test.py:659]: wrap properly

File: test/integration_test.py
Comment (hannes-ucsc) [inline test/integration_test.py:610]: why did you convert these to attributes?

File: test/integration_test.py
Comment (hannes-ucsc) [inline test/integration_test.py:646]: please move all cosmetic changes to a separate commit.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
> Integration test passes once #2399 is applied on top of this, otherwise the relevant test is skipped.

Reorder chain in that case. I'd rather retry the IT job after a sporadic fail than merge this blindly without validation.

I prefer not to see fixups from peer reviews.

File: test/integration_test.py
Comment (noah-aviel-dove) [reply test/integration_test.py:646]: https://github.com/DataBiosphere/azul/pull/2602/commits/9ac3ab01c9d3332ccba9fcfea5f64f19dda4107c

File: test/integration_test.py
Comment (noah-aviel-dove) [reply test/integration_test.py]: I don't remember my reasoning and no longer see a need for this change.

**Review by hannes-ucsc — COMMENTED** [review]
I've reviewed the PR this one is chained on. Please hold off requesting re-review until that PR has been merged. No bump.

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
For demo, run test in personal deployment, cancel it (Ctrl-C, I guess) and show that database is left in original state.

**Comment by noah-aviel-dove** [comment]
> For demo, run test in personal deployment, cancel it (Ctrl-C, I guess) and show that database is left in original state.

Control-C doesn't work, and neither does stopping the test via the PyCharm UI. The teardown doesn't run.

As an alternative, I can add an `assert False` to the end of the test and observe a successful reset.

**Comment by jessebrennan** [comment]
> Control-C doesn't work, and neither does stopping the test via the PyCharm UI. The teardown doesn't run.

Something is definitely wrong, because this directly contradicts the spec https://github.com/DataBiosphere/azul/blob/7a866cb7729c126468022a14b16a50570cf2b1d9/test/azul_test_case.py#L92.

I was suspect the inheritance hierarchy could be the problem, depending on where the execution is when you press Ctrl-C.

**Comment by hannes-ucsc** [comment]
You could put a sleep(10.0) in the test body to be sure the Ctrl-C is received in the right place.

**Comment by hannes-ucsc** [comment]
But `assert False` also works for the demo. The 10s sleep experiment could be used to confirm @jessebrennan's suspicion that there is a bug in AlwaysTearDownTestCase.

**Comment by noah-aviel-dove** [comment]
When the interrupt is caught during a call to `time.sleep`, the teardown runs successfully and the database is restored to its original state. I tried interrupting at multiple points in the test using `sleep` and none reproduced the teardown failure. 

Furthermore, I uncovered an error in my previous comment. I claimed that the teardown didn't run when interrupting the test, which I assumed because the database wasn't being reset. But I added some logging statements and it turns out it _is_ being run. The sequence of events is as follows:

1. The test starts.
1. The test is interrupted from the command line (no call to `time.sleep`).
2. The teardown starts.
3. The teardown finishes. The database is presumably reset.
4. The tracktrace is printed for `KeyboardInterrupt`.
5. The threads spawned by the test continue to execute, modifying the database.
6. The process exits.

**Pull request merged**
